optax/_src/loss.py

```diff
@@ -252,3 +252,99 @@ def log_cosh(
   errors = (predictions - targets) if (targets is not None) else predictions
   # log(cosh(x)) = log((exp(x) + exp(-x))/2) = log(exp(x) + exp(-x)) - log(2)
   return jnp.logaddexp(errors, -errors) - jnp.log(2.0).astype(errors.dtype)
+
+
+def ctc_loss(logits: chex.Array,
+             logitpaddings: chex.Array,
+             labels: chex.Array,
+             labelpaddings: chex.Array,
+             blank_id: int = 0,
+             logepsilon: float = -1e5) -> chex.Array:
+  """Computes CTC loss.
+
+  This function performs forward computation over an FSA with `N * 2` states
+  where `N` is the max number of labels. The states are split into two groups:
+  Phi states and emission states. a phi-state accepts repetition of
+  phi (blank)-symbols and transits to emission state when the correct label is
+  observed. An emission state accepts repetition of the label and transits to
+  the next phi states at any time (so called epsilon-transition).
+  Below, `B` denotes the batch size, `T` denotes the time steps in `logits`,
+  and `N` denotes the time steps in `labels`.
+
+  Args:
+    logits: (B, T, K)-array containing log-probabilities of each class.
+    logitpaddings: (B, T)-array. Padding indicators for `logits`.
+    labels: (B, N)-array containing reference integer labels.
+    labelpaddings: (B, N)-array. Padding indicators for `labels`. Currently,
+      `labels` must be right-padded, i.e. each row of `labelpaddings` must be
+      repetition of zeroes, followed by repetition of ones.
+    blank_id: Id for blank token.
+
+  Returns:
+    (B,)-array containing loss values for each sequence in the batch.
+  """
+  batchsize, unused_maxinputlen, num_classes = logits.shape
```

Comment:
Consider calling `chex.assert_rank(logits, 3)` for more informative exceptions

Reply:
Thanks, done.